# Calidad de los Datos.

Despues de descargar y preparar los datos. Procedemos a verificar la calidad de los datos. En esta libreta revisaremos las dimensiones de la calidad de datos.

* Integridad.
* Validez.
* Unicidad.
* Consistencia.
* Actualidad.
* Precisión.

## Origen de Datos

Actualmente poseemos dos fuentes de datos.
1. DGSIAP.
2. Data Mexico.

En esta libreta trabajaremos con el DGSIAP. Pero en el analisis final debemos aplicar todo lo revisado en ambas. 

In [2]:
from scripts.download import agro_project

# Obtenemos los datos establecidos en nuestro analisis.
file_tree = agro_project()


Conexion Establecida
Usted escogio Agricola
Los datos historicos se presentan en el espacio denomiado: Información por nivel
Los datos se encuentran divididos en 3 bloques:

1. Municipal
2. Nacional
3. No seguimiento


Obteniendo los datos en formato de arbol. Ahora procedemos con la descarga.

In [3]:
from scripts.download import load_filetree
# Descargamos los archivos de la estructura de directorios obtenida.
load_filetree(file_tree)

https://nube.agricultura.gob.mx/index.php?view=5D24F995-E734D711-1EA6F8D2-9D1387D7 Diccionario_agricola_2003_a_2025.xlsx /home/david/Documents/Repositories/academica/proyecto_agronomia/data/raw/dgsiap /home/david/Documents/Repositories/academica/proyecto_agronomia/data/raw/dgsiap/Diccionario_agricola_2003_a_2025.xlsx
El archivo ya fue descargado. Se encuentra en /home/david/Documents/Repositories/academica/proyecto_agronomia/data/raw/dgsiap/Diccionario_agricola_2003_a_2025.xlsx
https://nube.agricultura.gob.mx/index.php?view=10AE434F-A2158368-A120BC5A-EDF4AFAA&ANIO=2025 Cierre_agricola_mun_2025.csv /home/david/Documents/Repositories/academica/proyecto_agronomia/data/raw/dgsiap /home/david/Documents/Repositories/academica/proyecto_agronomia/data/raw/dgsiap/Cierre_agricola_mun_2025.csv
El archivo ya fue descargado. Se encuentra en /home/david/Documents/Repositories/academica/proyecto_agronomia/data/raw/dgsiap/Cierre_agricola_mun_2025.csv
https://nube.agricultura.gob.mx/index.php?view=10AE

### Formateo

Seguiremos con el formateo de los datos. Por el lado de los diccionarios debemos extraer la información relevante en forma de tabla. Mientras que en el caso de los conjuntos de datos debemos resolver el problema de encoding para perder la minima información posible.

In [4]:
from scripts.preparation import format_block, encode_block

blocks= list(file_tree.keys())
encode_project = {b: None for b in blocks}
# Calculando los encodings de los archivos de cada bloque de datos.
encode_project[blocks[0]] = encode_block(file_tree[blocks[0]])
print(encode_project)


{'Municipal': {'Codebook': <scripts.File.File object at 0x7f85012a0ec0>, 'Data': {'2025': {'file': <scripts.File.File object at 0x7f850189c550>, 'encoding': 'ISO-8859-1'}, '2024': {'file': <scripts.File.File object at 0x7f850189c410>, 'encoding': 'ISO-8859-1'}, '2023': {'file': <scripts.File.File object at 0x7f8501259cd0>, 'encoding': 'ISO-8859-1'}, '2022': {'file': <scripts.File.File object at 0x7f8501259f30>, 'encoding': 'ISO-8859-1'}, '2021': {'file': <scripts.File.File object at 0x7f85013efad0>, 'encoding': 'ISO-8859-1'}, '2020': {'file': <scripts.File.File object at 0x7f8500b186b0>, 'encoding': 'ISO-8859-1'}, '2019': {'file': <scripts.File.File object at 0x7f8500b187c0>, 'encoding': 'ISO-8859-1'}, '2018': {'file': <scripts.File.File object at 0x7f850129a050>, 'encoding': 'ISO-8859-1'}, '2017': {'file': <scripts.File.File object at 0x7f8500bdf150>, 'encoding': 'ISO-8859-1'}, '2016': {'file': <scripts.File.File object at 0x7f8501893200>, 'encoding': 'ISO-8859-1'}, '2015': {'file': <

Terminando el encoding de los datos procedemos a 
1. Guardar los encodings por si ocupamos volver a descargar los datos no tener que volver a detectar los encodings.
2. Generar los archivos para la calidad de datos.

In [5]:
# Guardar diccionario de encodings en un archivo JSON.
import json
from scripts.config import get_project_root


format_project = {b: None for b in blocks}
format_project[blocks[0]] = format_block(encode_project[blocks[0]], blocks[0])

{'2025': {'file': <scripts.File.File object at 0x7f850189c550>, 'encoding': 'ISO-8859-1'}, '2024': {'file': <scripts.File.File object at 0x7f850189c410>, 'encoding': 'ISO-8859-1'}, '2023': {'file': <scripts.File.File object at 0x7f8501259cd0>, 'encoding': 'ISO-8859-1'}, '2022': {'file': <scripts.File.File object at 0x7f8501259f30>, 'encoding': 'ISO-8859-1'}, '2021': {'file': <scripts.File.File object at 0x7f85013efad0>, 'encoding': 'ISO-8859-1'}, '2020': {'file': <scripts.File.File object at 0x7f8500b186b0>, 'encoding': 'ISO-8859-1'}, '2019': {'file': <scripts.File.File object at 0x7f8500b187c0>, 'encoding': 'ISO-8859-1'}, '2018': {'file': <scripts.File.File object at 0x7f850129a050>, 'encoding': 'ISO-8859-1'}, '2017': {'file': <scripts.File.File object at 0x7f8500bdf150>, 'encoding': 'ISO-8859-1'}, '2016': {'file': <scripts.File.File object at 0x7f8501893200>, 'encoding': 'ISO-8859-1'}, '2015': {'file': <scripts.File.File object at 0x7f850132c5f0>, 'encoding': 'ISO-8859-1'}, '2014': {

In [6]:
import pickle
# Guardar diccionario de encodings en un archivo pickle.
with open(get_project_root() / "docs" / "encode_project.pkl", 'wb') as f:
    pickle.dump(encode_project, f)

# Vemos los archivos formateados del proyecto.
format_project[blocks[0]]

{'Codebook': {'Diccionario': <scripts.File.File at 0x7f849c575c70>,
  'Variables': <scripts.File.File at 0x7f849c1564e0>},
 'Data': {'2025': <scripts.File.File at 0x7f849c1572f0>,
  '2024': <scripts.File.File at 0x7f85013c2f00>,
  '2023': <scripts.File.File at 0x7f85013c3020>,
  '2022': <scripts.File.File at 0x7f85013c2f90>,
  '2021': <scripts.File.File at 0x7f85013c2d50>,
  '2020': <scripts.File.File at 0x7f85013c2de0>,
  '2019': <scripts.File.File at 0x7f85013c2e70>,
  '2018': <scripts.File.File at 0x7f85013c2cc0>,
  '2017': <scripts.File.File at 0x7f849c156ba0>,
  '2016': <scripts.File.File at 0x7f849c156720>,
  '2015': <scripts.File.File at 0x7f849c1571d0>,
  '2014': <scripts.File.File at 0x7f849c157770>,
  '2013': <scripts.File.File at 0x7f849c1575c0>,
  '2012': <scripts.File.File at 0x7f849c1574a0>,
  '2011': <scripts.File.File at 0x7f849c155490>,
  '2010': <scripts.File.File at 0x7f849c1555b0>,
  '2009': <scripts.File.File at 0x7f849c157410>,
  '2008': <scripts.File.File at 0x7f

Ahora procedemos con los aspectos de la calidad de datos. 

## Verificación de Datos

En el resto de la libreta iremos viendo cada una de las dimensiones de los datos. En la libreta trabajaremos con un solo archivo pero esto se deberá aplicar para cada uno de las datos a usar para el futuro analisis.

Cargamos los datos previamente formateados. Tomaremos el Codebook asociado y el archivo más reciente.

In [42]:
import pandas as pd


agro_data = format_project[blocks[0]]
codebook = agro_data['Codebook']

# Cargar el diccionario de datos y sus variables.
catalog = pd.read_excel(codebook['Diccionario'].ubication, index_col=0)
datatypes = pd.read_excel(codebook['Variables'].ubication, index_col=0)

print(catalog.head())
print(datatypes.head())


                 Tipo  Long. Rangos\nválidos  \
Nombre\nde campo                               
Anio                N      4          {CCCC}   
Idestado            N      2       {1,…,32 }   
Nomestado           T     60           {CCC}   
Idddr               N      3       {1,…,193}   
Nomddr              T     60           {CCC}   

                                                        Descripción  \
Nombre\nde campo                                                      
Anio                                          Año de la información   
Idestado          Código que identifica a la entidad federativa ...   
Nomestado          Nombre oficial de la entidad federativa o estado   
Idddr             Código que identifica  al Distrito de Desarrol...   
Nomddr            Nombre que identifica  al Distrito de Desarrol...   

                                                      Observaciones  
Nombre\nde campo                                                     
Anio              Corresp

Notemos que el Diccionario de Datos se compone del catalogo de variables y los tipos de variables definidas. Por ello a lo que usualmente conocemos como diccionario de datos lo estaremos referenciando como catalogo.

En la descarga de datos podemos ver de forma más clara la importancia de la clase File y todo la construcción previa. Solo se hace una vez y facilita los analisis futuros.

In [43]:
# Cargamos la producción elf.link = pathmunicipal de 2025.
data = agro_data['Data']['2025']

# Producción Agricola Municipal 2025
## (Esto se puede deducir siguiendo la construcción del file tree)

### Leemos el archivo
prod_agricola = pd.read_excel(data.ubication)
prod_agricola.head()

,Anio,Idestado,Nomestado,Idddr,Nomddr,Idcader,Nomcader,Idmunicipio,Nommunicipio,Idciclo,...,Nomunidad,Idcultivo,Nomcultivo,Sembrada,Cosechada,Siniestrada,Volumenproduccion,Rendimiento,Preciomediorural,Valorproduccion
0,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,...,Tonelada,5490000,Avena forrajera en verde,1500.0,1500.0,0.0,35250.0,23.50,654.00,23053500.00
1,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,...,Tonelada,5670000,Brócoli,50.0,50.0,0.0,850.0,17.00,6887.78,5854613.00
2,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,2,...,Tonelada,5490000,Avena forrajera en verde,165.0,165.0,0.0,3700.0,22.42,645.00,2386500.00
3,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,2,...,Tonelada,5670000,Brócoli,60.0,60.0,0.0,785.2,13.09,5161.69,4052958.99
4,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,2,...,Tonelada,5740000,Calabacita,30.0,30.0,0.0,490.0,16.33,5285.21,2589752.90


Finalmente, un adelanto del espacio de la validez.

In [125]:
# Esta lista puede modificarse en función de las necesidades
python_types = ["str", "int", "float", "float"]

project_datatypes = {
    t: python_types[k] 
    for k, t in enumerate(datatypes.index)}
    
print(project_datatypes)

{'T': 'str', 'N': 'int', 'D': 'float', '#': 'float'}


### Integridad

Pregunta Clave: ¿Los datos están completos y no faltan valores?

Para este paso nos centraremos en dos elementos:
* Coherencia de los datos.
* Datos faltantes o Nulos

#### Coherencia.

Descargamos diccionarios de datos. Esperamos que las columnas coincidan con las variables definidas.

In [61]:
catalog_variables = list(catalog.index)
agro_columns = prod_agricola.columns.to_list()

print("Verifcamos mismo tamaño de columnas", len(catalog_variables) == len(agro_columns))
print("Verificamos mismos elementos", set(catalog_variables) == set(agro_columns))

Verifcamos mismo tamaño de columnas True
Verificamos mismos elementos False


Son el mismo número de elementos. Entonces alguno de las columnas esta escrita distinto.

In [62]:

diff = [(ref, data)
        for ref, data in zip(catalog_variables, agro_columns)
        if ref != data]
print("Columnas distintas")
print(diff)

Columnas distintas
[('Precio', 'Preciomediorural')]


Entnces la columna de Preciomediorural tomo el lugar de la variable Precio en el diccionario. Revisando el diccionario vemos que es la variable precio. 

In [63]:
desc= catalog.loc['Precio']['Descripción']
obsv = catalog.loc['Precio']['Observaciones']
desc, obsv

('Precio medio rural',
 'Precio medio rural, la unidad de medida son pesos por tonelada, con excepción de los cultivos que tienen otra métrica la cual se señala en el nombre del cultivo (gruesa, manojo, planta, entre otros).')

Entonces vemos que el nombre es distinto. Pero la descripción de la columna Precio coincide con el nombre de la columna problemática. Entonces podemos reemplazar las columnas del conjunto de datos con la del diccionario para obtener coherencia entre ambas. 

In [ ]:
# Ultima verificación antes de hacer el cambio
for ref, data in zip(dict_variables, agro_columns):
    print(ref, data)

prod_agricola.columns = catalog_variables

Anio Anio
Idestado Idestado
Nomestado Nomestado
Idddr Idddr
Nomddr Nomddr
Idcader Idcader
Nomcader Nomcader
Idmunicipio Idmunicipio
Nommunicipio Nommunicipio
Idciclo Idciclo
Nomcicloproductivo Nomcicloproductivo
Idmodalidad Idmodalidad
Nommodalidad Nommodalidad
Idunidadmedida Idunidadmedida
Nomunidad Nomunidad
Idcultivo Idcultivo
Nomcultivo Nomcultivo
Sembrada Sembrada
Cosechada Cosechada
Siniestrada Siniestrada
Volumenproduccion Volumenproduccion
Rendimiento Rendimiento
Precio Preciomediorural
Valorproduccion Valorproduccion
       Anio  Idestado       Nomestado  Idddr          Nomddr  Idcader  \
0      2025         1  Aguascalientes      1  Aguascalientes        1   
1      2025         1  Aguascalientes      1  Aguascalientes        1   
2      2025         1  Aguascalientes      1  Aguascalientes        1   
3      2025         1  Aguascalientes      1  Aguascalientes        1   
4      2025         1  Aguascalientes      1  Aguascalientes        1   
...     ...       ...         

Ahora podemos verificar que coinciden, debemos de ver la lista vacia. 

In [76]:
def agro_help(var, guide=catalog):
    assert var in guide.index, "Revisa el nombre de la variable"
    desc= guide.loc[var]['Descripción']
    obsv = guide.loc[var]['Observaciones']
    return {var:{
        "Descripcion": desc,
        "Observaciones": obsv
    }}

agro_columns = prod_agricola.columns.to_list()
diff = [(ref, data)
        for ref, data in zip(catalog_variables, agro_columns)
        if ref != data]
print(diff)

[]


Con los nombres de las columnas adaptados al diccionario (Como un diccionario de datos es inmutable, procuramos que las columnas coincidan con el diccionario, no alrevés) podemos continuar con la segunda parte. 

#### Valores Nulos.
Dentro de la parte de la integridad esta el verificar la existencia de los datos nulos. Puesto que debemos indagar lo suficiente para saber su motivo y como proceder al respecto. Entonces primero debemos ubicarlos.

In [67]:
# Revisamos las columnas
columns_na = prod_agricola.columns[prod_agricola.isna().any()].tolist()
print(columns_na)

['Rendimiento', 'Precio']


Vemos que todos los datos nulos se encuentran en dos columnas. Entonces vamos a centrarnos en esos.

In [69]:
prod_na = prod_agricola[prod_agricola.isna().any(axis=1)]
prod_na.shape

(415, 24)

Vemos que poseemos 415 entradas con valores nulos en una o en las dos columnas previamente revisadas.

In [72]:
prod_na.head(n=8)

,Anio,Idestado,Nomestado,Idddr,Nomddr,Idcader,Nomcader,Idmunicipio,Nommunicipio,Idciclo,...,Nomunidad,Idcultivo,Nomcultivo,Sembrada,Cosechada,Siniestrada,Volumenproduccion,Rendimiento,Precio,Valorproduccion
28,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,3,...,Tonelada,5050000,Agave,8.0,0.0,0.0,0.0,NaN,NaN,0.0
111,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,11,San Francisco de Los Romo,3,...,Tonelada,7960000,Nuez,79.0,0.0,0.0,0.0,NaN,NaN,0.0
344,2025,2,Baja California,2,Ensenada,1,Tijuana,5,Playas de Rosarito,3,...,Tonelada,9130000,Uva,4.0,0.0,4.0,0.0,NaN,NaN,0.0
367,2025,2,Baja California,2,Ensenada,2,Tecate,3,Tecate,3,...,Planta,8060000,Palma de ornato (planta),2.0,0.0,2.0,0.0,NaN,NaN,0.0
384,2025,2,Baja California,2,Ensenada,3,Ensenada,1,Ensenada,1,...,Tonelada,5480000,Avena forrajera achicalada,4.0,0.0,4.0,0.0,NaN,NaN,0.0
385,2025,2,Baja California,2,Ensenada,3,Ensenada,1,Ensenada,1,...,Tonelada,5510000,Avena grano,1900.0,0.0,1900.0,0.0,NaN,NaN,0.0
386,2025,2,Baja California,2,Ensenada,3,Ensenada,1,Ensenada,1,...,Tonelada,5890000,Cebada forrajera achicalada,15.0,0.0,15.0,0.0,NaN,NaN,0.0
387,2025,2,Baja California,2,Ensenada,3,Ensenada,1,Ensenada,1,...,Tonelada,5920000,Cebada grano,1800.0,0.0,1800.0,0.0,NaN,NaN,0.0


Revisando las primeras columnas vemos algo interesante. Toda la muestra que no posee Rendimiento ni Precio tienen Cosechada y Volumenproduccion en 0.0. El contexto puede darnos una idea del motivo. Pero siempre debemos consultar el diccionario de datos. 

In [77]:
print(agro_help("Sembrada"))
print(agro_help("Cosechada"))
print(agro_help("Siniestrada"))
print(agro_help("Volumenproduccion"))
print(agro_help("Precio"))
print(agro_help("Rendimiento"))
print(agro_help("Valorproduccion"))

{'Sembrada': {'Descripcion': 'Superficie sembrada del cultivo', 'Observaciones': 'La unidad de medida son hectáreas. Es la superficie agrícola en la cual se deposita la semilla de cualquier cultivo, previa preparación del suelo.'}}
{'Cosechada': {'Descripcion': 'Superficie cosechada del cultivo', 'Observaciones': 'La unidad de medida son hectáreas. Es la superficie de la cual se obtuvo producción del cultivo indicado.'}}
{'Siniestrada': {'Descripcion': 'Superficie siniestrada del cultivo', 'Observaciones': 'La unidad de medida son hectáreas. Es la superficie de la cual no se obtuvo producción debido a que se perdió por alguna causa, solo aplica para cultivos anuales, ya que los perennes se pueden recuperar las plantaciones en el año siguiente.'}}
{'Volumenproduccion': {'Descripcion': 'Volumen de producción', 'Observaciones': 'Volumen de producción de la superficie cosechada cuya unidad de medida son las toneladas, con excepción del maguey pulquero y trigo ornamental, con una métrica de

Consultando el diccionario podemos formular una mejor hipotesis para la existencia de los datos. 

* Como el rendimiento se mide con tonelada por hectaria. Podemos deducir que es el cociente de Volumenproduccion y Sembrada. Como ambos son cero obtenemos NaN.
* También, podemos deducir que si Cosechada es 0.0 no se pudo vender y por lo tanto no se registro Precio.

Estas deducciones no tienen porque ser el motivo real. Pero serán hipotesis de las partiremos para el resto del analisis. Pero hay un detalle. ¿Que pasá con el resto de registros?, ¿Cumplen nuestras hipotesis?. Revisemos

In [ ]:
# Verificamos si existe registro cuando Volumenproducción no es cero.
check = prod_na[prod_na['Volumenproduccion']!=0.0]
print(check)

Empty DataFrame
Columns: [Anio, Idestado, Nomestado, Idddr, Nomddr, Idcader, Nomcader, Idmunicipio, Nommunicipio, Idciclo, Nomcicloproductivo, Idmodalidad, Nommodalidad, Idunidadmedida, Nomunidad, Idcultivo, Nomcultivo, Sembrada, Cosechada, Siniestrada, Volumenproduccion, Rendimiento, Precio, Valorproduccion]
Index: []

[0 rows x 24 columns]


Vemos que nos devolvio una lista vaica. Entonces todos los valores que posen NaN en Precio y/o Rendimiento poseen Volumenproduccion igual a 0.0. Por lo tanto, nuestras hipotesis se cumplen para todo el conjunto de datos problematico.

**Veredicto** En este caso, bajo nuestras hipotesis, los valores NaN poseen una explicación relevante dentro del conjunto de datos. Se mantendrán pero deben tomarse en cuenta para cualquier pregunta que nos hagamos en el futuro.

### Validez

Pregunta clave:
¿Los datos cumplen con las reglas y restricciones
definidas?

La coherencia entra dentro de este bloque. Sin embargo, se trabajo al comienzo para hacer más comodo el resto del proceso.

La validez se puede resumir en lo siguente que ya tiene el catalogo de datos. 

In [86]:
rules = catalog[['Tipo','Long.','Rangos\nválidos']]
rules

,Tipo,Long.,Rangos\nválidos
Nombre de campo,,,
Anio,N,4,{CCCC}
Idestado,N,2,"{1,…,32 }"
Nomestado,T,60,{CCC}
Idddr,N,3,"{1,…,193}"
Nomddr,T,60,{CCC}
Idcader,N,2,"{1,…,19}"
Nomcader,T,90,{CCC}
Idmunicipio,T,3,"{1,….,570}"
Nommunicipio,T,45,{CCC}


In [88]:
datatypes

,Variable
T,Texto
N,Numérico entero
D,Numérico decimal
#,Carácter numérico.


La fuente, el DGSIAP, definió las reglas que cada variable debe seguir.

* Tipo de Dato.
* Longitud.
* Rangos Validos.

Entonces debemos verificar cada uno de puntos. Sin embargo.

In [97]:
# Tipo de Dato.
prod_dtypes = prod_agricola.dtypes
print(prod_dtypes)

Anio                    int64
Idestado                int64
Nomestado                 str
Idddr                   int64
Nomddr                    str
Idcader                 int64
Nomcader                  str
Idmunicipio             int64
Nommunicipio              str
Idciclo                 int64
Nomcicloproductivo        str
Idmodalidad             int64
Nommodalidad              str
Idunidadmedida          int64
Nomunidad                 str
Idcultivo               int64
Nomcultivo                str
Sembrada              float64
Cosechada             float64
Siniestrada           float64
Volumenproduccion     float64
Rendimiento           float64
Precio                float64
Valorproduccion       float64
dtype: object


En general no nos interesa la precisión del entero y/o el flotante (el número despues de int/float). Entonces solo revisaremos que forme parte de los tipos de dato elemental.

In [126]:
for col in agro_columns:
    col_type = rules.loc[col, 'Tipo']
    col_dtype =str(project_datatypes[col_type])
    ac_dtype = str(prod_dtypes[col])
    try:
        ac_dtype.index(col_dtype)
        project_datatypes[col_type] = prod_dtypes[col]
    except:
        print(col, ac_dtype, col_dtype)

Idmunicipio int64 str


Como podemos ver la columna Idmunicipio viene definida como un entero pero el diccionario de datos lo marco como texto (T). Pero se nos presenta un dilema. 

In [127]:
rules.loc[[r for r in rules.index if r[:2] == "Id"]]

,Tipo,Long.,Rangos\nválidos
Nombre de campo,,,
Idestado,N,2,"{1,…,32 }"
Idddr,N,3,"{1,…,193}"
Idcader,N,2,"{1,…,19}"
Idmunicipio,T,3,"{1,….,570}"
Idciclo,N,1,"{1,….,3}"
Idmodalidad,N,1,"{1,2}"
Idunidadmedida,N,1,"{1,….,6}"
Idcultivo,N,3,"{1,….,410}"


Vemos que Idmunicipio es el único identificador que es texto, el resto son enteros. 

**Dilema** Hace un momento comentamos que el diccionario es inmutable. Sin embargo, considerando que Idestado es un entero, podemos deducir que no hay un argumento fuerte para que sea un elemento de texto y habra sido un error de dedo.

**Veredicto** No vamos a modificar el diccionario de datos. Sin embargo, tomando en cuenta que para el analisis posterior convertiremos los datos Id enteros a categoricos no afecta al analisis. Pero esto debe ser mencionado en el resumen de la calidad de los datos.

Con los tipos de datos corroborado ahora procedemos a la parte de los rangos. Para esto lo haremos de forma natural y luego probaremos con la libreria pandatic.

In [ ]:
project_datatypes

{'T': <StringDtype(storage='python', na_value=nan)>,
 'N': dtype('int64'),
 'D': dtype('float64'),
 '#': 'float'}

In [ ]:
rl_mun = rules.loc['Idmunicipio']
long = rl_mun['Long.']
range_mun = rl_mun["Rangos\nválidos"]

def long_rule(val, long):
    assert str(val), "Revisa la entrada"
    check = 1 if len(str(val)) <= long else 0
    return check

def range_rule(val, rage: tuple):
    a, b = rage
    check_set = set(range(a, b+1))
    return val in check_set

def interval_rule(val, interval: tuple, config = (True, True)):
    a, b = interval
    value = -1 
    if config[0]:
        if config[1]:
            value= a <= val <= b
        else:
            value= a <= val < b
    else:
        if config[1]:
            value= a < val <= b
        else:
            value= a < val < b
    return value

print(rl_mun, long, rl_mun)




g = rules[rules['Tipo']=="N"]
import re


for gi in g.index:
    k = re.search(r"{(.*)}",g.loc[gi, "Rangos\nválidos"])
    ranges = k.group(1)
    g_min = ranges.split(",")[0]
    g_max = ranges.split(",")[-1]
    print(ranges, g_min, g_max)


Tipo                        T
Long.                       3
Rangos\nválidos    {1,….,570}
Name: Idmunicipio, dtype: object 3 Tipo                        T
Long.                       3
Rangos\nválidos    {1,….,570}
Name: Idmunicipio, dtype: object
CCCC CCCC CCCC
1,…,32  1 32 
1,…,193 1 193
1,…,19 1 19
1,….,3 1 3
1,2 1 2
1,….,6 1 6
1,….,410 1 410


In [ ]:
m = pd.DataFrame(index=[0], columns=["A", "B"], data=0, dtype=pd.Int64Dtype)


print(m.dtypes)
print(int(m.loc[0, 'A']))


A    Int64
B    Int64
dtype: object
0


/tmp/ipykernel_76514/767915425.py:1: UserWarning: Instantiating Int64Dtype without any arguments.Pass a Int64Dtype instance to silence this warning.
  m = pd.DataFrame(index=[0], columns=["A", "B"], data=0, dtype=pd.Int64Dtype)


### Unicidad 

Pregunta clave
¿Hay datos duplicados?

### Consistencia

Pregunta clave
¿Los datos son coherentes entre diferentes fuentes
y representaciones?

### Actualidad
¿Los datos están actualizados y reflejan la realidad
actual?

### Precisión

¿Los datos son correctos y libres de errores?